src/distibutions/base-distribution.ts

```diff
@@ -0,0 +1,253 @@
+import * as assert from 'assert';
+
+import * as path from 'path';
+import os from 'os';
+import fs from 'fs';
+
+import {INodejs, INodeVersion, INodeVersionInfo} from './base-models';
+
+export default abstract class BaseDistribution {
+  protected httpClient: hc.HttpClient;
+  protected osPlat = os.platform();
+
+  constructor(protected nodeInfo: INodejs) {
+    this.httpClient = new hc.HttpClient('setup-node', [], {
+      allowRetries: true,
+      maxRetries: 3
+    });
+  }
+
+  protected abstract getDistributionUrl(): string;
+  protected abstract evaluateVersions(nodeVersions: string[]): string;
+
+  public async getNodeJsInfo() {
+    let nodeVersions: INodeVersion[] | undefined;
+    if (this.nodeInfo.checkLatest) {
+      nodeVersions = await this.getNodejsVersions();
+      const versions = this.filterVersions(nodeVersions);
+      const evaluatedVersion = this.evaluateVersions(versions);
+
+      if (evaluatedVersion) {
+        this.nodeInfo.versionSpec = evaluatedVersion;
+      }
+    }
+
+    let toolPath = this.findVersionInHoostedToolCacheDirectory();
+    if (toolPath) {
+      core.info(`Found in cache @ ${toolPath}`);
+    } else {
+      nodeVersions = nodeVersions ?? (await this.getNodejsVersions());
+      const versions = this.filterVersions(nodeVersions);
+      const evaluatedVersion = this.evaluateVersions(versions);
+      if (!evaluatedVersion) {
+        throw new Error(
+          `Unable to find Node version '${this.nodeInfo.versionSpec}' for platform ${this.osPlat} and architecture ${this.nodeInfo.arch}.`
+        );
+      }
+      const toolName = this.getNodejsDistInfo(evaluatedVersion, this.osPlat);
+      toolPath = await this.downloadNodejs(toolName);
+    }
+
+    if (this.osPlat != 'win32') {
+      toolPath = path.join(toolPath, 'bin');
+    }
+
+    core.addPath(toolPath);
+  }
+
+  protected findVersionInHoostedToolCacheDirectory() {
+    return tc.find('node', this.nodeInfo.versionSpec, this.nodeInfo.arch);
+  }
+
+  protected async getNodejsVersions(): Promise<INodeVersion[]> {
+    const initialUrl = this.getDistributionUrl();
+    const dataUrl = `${initialUrl}/index.json`;
+
+    let response = await this.httpClient.getJson<INodeVersion[]>(dataUrl);
+    return response.result || [];
+  }
+
+  protected getNodejsDistInfo(version: string, osPlat: string) {
+    let osArch: string = this.translateArchToDistUrl(this.nodeInfo.arch);
+    version = semver.clean(version) || '';
+    let fileName: string =
+      osPlat == 'win32'
+        ? `node-v${version}-win-${osArch}`
+        : `node-v${version}-${osPlat}-${osArch}`;
+    let urlFileName: string =
+      osPlat == 'win32' ? `${fileName}.7z` : `${fileName}.tar.gz`;
+    const initialUrl = this.getDistributionUrl();
+    const url = `${initialUrl}/v${version}/${urlFileName}`;
+
+    return <INodeVersionInfo>{
+      downloadUrl: url,
+      resolvedVersion: version,
+      arch: osArch, // have to be arch but not osArch,
```

Comment:
I do not understand why this comment is needed and what it is about

Reply:
Done. Thank you.